contents/swift/welcome-to-swift.md

```diff
@@ -0,0 +1,144 @@
+println("\(x.data), \(y.data)") // prints "42, 42"
+```
+
+You can think of Structs as a way to create instances that are immutable since no other part of your code can change it, which can help to make things a lot easier.
+
+If you wish to find out more but pressed for time, here is an [article](https://cocoacasts.com/value-types-and-reference-types-in-swift/) that sumamrises the difference between the 2 types, as well as the benefits of value types and when to use them  
+
+Read [this](https://medium.com/capital-one-developers/reference-and-value-types-in-swift-de792db330b2) if you are looking for something more verbose which explains the difference between the 2 types with diagrams.
+
+
+## Protocol Oriented Programming
+
+The heart of Swift is Protocol Oriented Programming(POP) which is about abstraction and simplicity. POP helps to solve the [bloat that is sometimes caused by Object Oriented Programming(OOP)](http://blogs.perl.org/users/sid_burn/2014/03/inheritance-is-bad-code-reuse-part-1.html). If you ever find yourself having to inherit from multiple classes, you probably should consider using protocols instead.
+
+Here's some code to serve as a brief introduction to POP
+
+We first introduce our protocols.
+
+```
+protocol Bird {
+  var name: String { get }
+  var canFly: Bool { get }
+}
+ 
+protocol Flyable {
+  var airspeedVelocity: Double { get }
+}
+```
+
+Next, we introduce the structs that confrom to the protocols above.
+
+```
+// Penguins can't fly ):
+struct Penguin: Bird {
+  let name: String
+  let canFly = false
+}
+
+struct Eagle: Bird, Flyable {
+  let name: String
+  let canFly = true
+ 
+  var airspeedVelocity: Double {
+    return 160.0
+  }
+}
+```
+
+And if you haven't noticed, protocols are extremely similar to interfaces in Java. 
+
+To understand more about POP, I highly recommend watching this [2015 WWDC talk](https://www.youtube.com/watch?v=g2LwFZatfTI).
+
+## Automatic Reference Counter
+
+A few keywords special to Swift are `strong`, `weak` and `unowned`, which have to do with Swift's way of memory management, [Automatic Reference Counting (ARC)](https://developer.apple.com/library/content/documentation/Swift/Conceptual/Swift_Programming_Language/AutomaticReferenceCounting.html). Essentially, when an instance of class is created, a chunk of memory is allocated to it until it is no longer strongly referenced by anything. Reference are strong by default. Thus, if we have an Object A (a `UIViewController`) that creates Object B (a `UIAlertController`), B would be strongly referenced by A. However, B might also need access to a variable in A, resulting in A being strongly referenced by B, creating a reference cycle. Reference cycles are bad, because they cause memory leaks. Even though A and B are no longer needed eventually, A and B will still sit in memory since they are both strongly referenced by each other. This is why we need the `strong`, `weak` and `unowned` keywords, to resolve reference cycles. 
```

Comment:
- `A few keywords`~`special`~`unique`
- `instance of a class`
- `References are`
- `creates an Object B`
- ~`resulting in`~`such that A`~`being`~`may be strongly referenced`
- ~`creating`~`resulting in a reference cycle`